Added Edit button to the DataForm panel layout

jimjasson opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR improves the edit interaction for DataForm panel fields by replacing the clickable field value with a separate edit button that appears on hover. 

## Why?

The previous implementation made entire field values look and behave like buttons. This created an accessibility issue where screen readers would only announce "Edit [field name]" without revealing the current field value, leaving users without crucial context.

## How?
- **Component changes**: Refactored `SummaryButton` to render a wrapper div containing the summary content and a separate icon button, replacing the previous single-button implementation.
- **Styling**: Added `.dataforms-layouts-panel__summary-wrapper` with hover states that fade the edit button in/out, plus `@media (hover: none)` to keep buttons always visible on touch devices.
- **Accessibility**: Added `useInstanceId` to generate unique IDs for summary content, then linked each edit button to its summary using `aria-describedby` so screen readers announce both the action and current value.

## Questions

- For `readOnly` fields, I decided to not show the Edit button. Do you agree with this decision?
- The `labelPosition` prop of `SummaryButton` was previously used to determine the `variant` of the button that overlayed the field value. I don't think that this is needed any longer. Do you agree?

## Testing Instructions

1. Run `npm run storybook:dev`.
2. Navigate to `DataForm` examples. 
3. Validate that the values of `panel` fields are no longer clickable -- instead an edit button appears on hover. 
4. Validate that the edit flow works as it used to. 
5. For read-only fields, validate that you don't see an edit button on hover. 
6. For mobile/touch devices, validate that the edit icon is always visible. 

### Testing Instructions for Keyboard

1. Run `npm run storybook:dev`.
2. Navigate to `DataForm` examples. 
3. Click on one of the examples that contain a `panel` field, e.g the `Customer` card. 
4. Use the `tab` key to navigate the page. 
5. Validate that the edit buttons appear when focused by the tab. 
7. Validate that by clicking `enter` the edit panel opens up. 


https://github.com/user-attachments/assets/da4a9f48-65d6-4ec2-a508-41f5a2251bda


## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
| <img width="1083" height="682" alt="Markup on 2025-11-06 at 10:27:19" src="https://github.com/user-attachments/assets/a0d8fb04-6b6f-418d-bd7f-6d9991389710" /> |<img width="1105" height="771" alt="Markup on 2025-11-06 at 10:27:41" src="https://github.com/user-attachments/assets/7b1561cd-9ced-465a-9ca5-724362597c2b" />|

**Desktop experience**

https://github.com/user-attachments/assets/62f9d33d-adf7-4145-a2f4-210f947a6be2

**Mobile experience**

https://github.com/user-attachments/assets/1965cb97-9284-47eb-aabd-db3bb49c1969

jimjasson committed:
Added Edit button to DataForm panels

jimjasson committed:
Updated E2E test

jimjasson committed:
More E2E tests updates

jimjasson committed:
Updated selector in E2E test

jimjasson committed:
Added changelog

jimjasson committed:
Merge branch 'trunk' into tweak/add-edit-button-to-dataform-panels

oandregal commented:
Design-wise, it is difficult to tell apart the label from the actual value. I'm going to summon design folks for this:

https://github.com/user-attachments/assets/ed370830-253c-4ed6-9a6a-1ac7cdc4cf19

Additionally, we need to figure out how custom Edit controls work in this scenario. This is how QuickEdit behaves in Site Editor > Pages (need to have enabled the experiment in Gutenberg > Experiments). Note how the [template field](https://github.com/jimjasson/gutenberg/blob/7b3a1ea55534b5968ca1131ece1fcc7b375bee75/packages/fields/src/fields/template/template-edit.tsx#L147)) renders its own dropdown:

https://github.com/user-attachments/assets/1908bdc9-971b-4b2b-9a0a-c269322907a7

jameskoster commented:
In terms of the general UX a dedicated edit button makes sense and seems like a nice a11y enhancement, but I think the entire field should also be clickable to trigger the modal/popover. Additionally the pencil button should appear when hovering the whole field to address Andrés first feedback. I made a prototype [here](https://incandescent-piroshki-40f129.netlify.app/) to demonstrate the interaction.

Otherwise @poligilad-auto's feedback sounds good. I could potentially see un-editable field _values_ being gray too to further distinguish, but it's not a strong feeling. Here's a diagram to illustrate in case it's useful:

<img width="929" height="762" alt="Screenshot 2025-11-21 at 16 51 48" src="https://github.com/user-attachments/assets/e990a90a-6642-42e7-b96a-ca95ede54e43" />

> Note how the template field renders its own dropdown with one or two actions before opening the modal.

Is this definitely something we want to support moving forwards? _Actions_ potentially feel a bit out of scope for DataForm and things can get a bit complex when actions are conditionally available (IE what do we do after the template has been reset, show a menu containing one item?). In this particular example swapping the template seems fine, but the 'reset' action should arguably live elsewhere, grouped with other contextual actions like "Rename", "Trash", and so on. 

If we _do_ want to support this then maybe we replace the pencil icon with a more icon and show it when hovering the field. But in this case only the icon button would be interactive, not the entire field.

<img width="769" height="332" alt="Screenshot 2025-11-21 at 17 03 00" src="https://github.com/user-attachments/assets/fc681111-a91d-4101-9603-48446da91a47" />

jimjasson commented:
Regarding empty buttons: 

> Maybe we can solve this with copy, like this. "Enter value" for example, with gray text to indicate that it's empty.

I think that it should be up to each field to decide how it wants to handle its empty state and DataForm shouldn't provide a default one. 

For example, for some fields it might be more appropriate to have a `-`, for others a placeholder text etc. This can be controlled via the field's `render` function. 

What do you think @oandregal ? Any chance that the issue with the empty value in the Storybook can be fixed by adding a render function to the Title field?

jimjasson committed:
- Added hover background and theme color to panel fields
- Enabled clicking entire field row to trigger edit action
- Revelaed edit button on row hover (previously only on value hover)
- Increased field value font-weight to 500 for better hierarchy
- Styled read-only fields with muted gray-700 color
- Implemented smart click detection to avoid triggering on interactive elements
- Added `is-read-only` class to read only fields
- Improved accessibility for read-only buttons

jameskoster commented:
@verofasulo do you mean something like this?

<img width="1190" height="914" alt="Screenshot 2025-12-19 at 15 37 57" src="https://github.com/user-attachments/assets/dd61010b-d3c9-481e-8923-1052fb7e9c85" />

I think the main feedback I received about this approach is that in 99% of cases "it's going to be a lot of pencils" 😅 I think that's valid given _most_ forms won't have many non-editable fields. We should optimise for the more common use cases.

> appear disabled

I mentioned this elsewhere, but non-editable fields basically _are_ disabled if we think about the form context.

I think the hierarchy point can potentially be addressed by grouping the fields based on whether they're editable or not. Obviously this is a consumer/implementation point though.

oandregal commented:
@jameskoster @verofasulo @poligilad-auto what are the next steps on this one?

The issue with empty values has been mitigated with some improvements, though it's still far from being good enough:

https://github.com/user-attachments/assets/d211eadb-0108-4cd6-93ac-7a2f76e31a59

From what we have now, these are the design changes I see discussed:

1. Make the whole field clickable, as opposed to only the field's value as of now.
2. Editable vs. read-only fields:
   1. icon is visible for editable fields at all times
   2. icon is visible for editable fields only on hover + read-only fields are gray-out

Can we ship 1 (making the entire field clickable) as a first step? That's a good improvement that appears to have agreement. In the meanwhile, design can iterate on the second point.

jameskoster commented:
> Can we ship 1 (making the entire field clickable) as a first step? That's a good improvement that appears to have agreement. In the meanwhile, design can iterate on the second point.

Yes I think that's the way to go. We're at a bit of an impasse with the second point... I don't think we should show the icon permanently as it's going to result in a lot of superfluous repetition in many cases. The document inspector for example:

<img width="738" height="410" alt="Screenshot 2026-01-09 at 13 34 45" src="https://github.com/user-attachments/assets/89831891-eb97-4f0c-9d27-e2fc235c331f" />
